runtime/onert/core/include/backend/train/ExtraTensorRequest.h

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2024 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __ONERT_BACKEND_EXTRA_TENSOR_REQUEST_H__
+#define __ONERT_BACKEND_EXTRA_TENSOR_REQUEST_H__
+
+#include "backend/train/ExtraTensor.h"
+
+namespace onert
+{
+namespace backend
+{
+namespace train
+{
+
+enum class ExtraTensorLifeTime
+{
+  BACKWARD,            // alive during backward()
+  FORWARD_TO_BACKWARD, // alive from forward to backward()
+};
```

Comment:
first of all, I may be asking the wrong question due to my lack of understanding of this function, so please understand. 

what does lifetime mean?
is it dynamically allocating/deallocating memory at the start/end of the section(backward, forward_to_backward)?
and, what does "Extra" mean in ExtraTensor?

Reply:
> first of all, I may be asking the wrong question due to my lack of understanding of this function, so please understand.

Thanks for your review and interest! I'm glad that I could explain it. 

> what does "Extra" mean in ExtraTensor? 

ExtraTensor means a tensor additionally requested while doing backward(backpropagte).
 
For example, while backwarding FullyConnected, we need to calculate X^T and W^T. 
Since X^T and W^T are 'additionaly' needed because of mathmetical properties of FC, I used the word 'Extra' to mark it additional tensor related to specific layer. 

<img src="https://github.com/user-attachments/assets/02023387-3106-454b-895c-a8ae02c134b5" width="40%">

Reply:
> what does lifetime mean? 

Lifetime means the time that additional tensors must remain alive.

In the example above, W^T and X^T of FullyConnected are calculated and used only during backpropagte. 
So, the lifetime of these tensors are limited to backward(). (mark it `BACKWARD`)  

Some Extra Tensors have to be alive while both forwarding and backwarding. 
So, I'd like to mark it as `FORWARD_TO_BACKWARD`. 

> is it dynamically allocating/deallocating memory at the start/end of the section(backward, forward_to_backward)?
 
With Lifetime, we can know in advance when and how much tensor will be needed.
So I'd like to pre-allocate maximum memory before the training process start.

Reply:
I think that my explanation is somehow verbose and a bit hard to understand :cry: 
If you're more interested I could explain the details offline!

Reply:
**1.**

When I suggested `LayerScopeTensor` in https://github.com/Samsung/ONE/pull/13605#issuecomment-2274478201 based on your comment

> ExtraTensor is a tensor that is accessed within one operation layer.
> In other words, the scope of the extra tensor is confined to one specific layer.

But it seems misleading to use the scope is one specific layer.
It may be accessed in the specific layer only, (but it is true for all other tensors) but its lifetime is longer than the specific layer.

I expected it is like a local variable, which is released when it is out of the scope.

**2.**

Why do we need `ExtraTensor` to get the scope of a Tensor? 
Is it possible to use use-def chain to get the scope of a Tensor?
It would be better not to introduce yet another tensor if possible.

Reply:
> Is it possible to use use-def chain to get the scope of a Tensor? 

As I understood, using the `use-def` chain is not easy in this problem. (at least for now) 

Current `use-def` chain has below structure.

```cpp
// UseDefChain.h

class UseDefChain
{
	const Operand &_operand; 
	std::set<TrainingOperationIndex> _uses;
	std::set<TrainingOperationIndex> _defs;
};
```
This assumes that `_operand` (tensor) is defined and used by **operations in the graph**.
But, ExtraTensors are not defined and used by operations in the graph.

For example, while backwarding... below computation will be done.
$W^T$ and $X$ are used by under the formular, but under formulars are not shown in the graph. 
 
$\frac{\partial L}{\partial X}$ = $\frac{\partial L}{\partial Z} W^T $

$\frac{\partial L}{\partial W} = (\frac{\partial L}{\partial Z})^T X$

Reply:
**about 1.**

> But it seems misleading to use the scope is one specific layer.  

I thought the word 'scope' meant 'where the tensor can be accessed' and it doesn't imply the lifetime.
So for me, the word 'scope' looks proper in this case. 


> It may be accessed in the specific layer only, (but it is true for all other tensors)
 
Aha, I agree with your point. 
About the weight tensors, it is accessed by one specific layer. 
I need to re-consider the Tensor naming and its property carefully.